File: frontend/js/fhq.frontend.lib.js
/*
  other implementation
*/

function FHQFrontEndLib() {

	// helpers function
	this.createUrlFromObj = function(obj) {
		var str = "";
		for (k in obj) {
			if (str.length > 0)
				str += "&";
			str += encodeURIComponent(k) + "=" + encodeURIComponent(obj[k]);
		}
		return str;
	}
	this.baseUrl = "http://fhq.keva.su/";
	this.token = "";
	this.client = "FHQFrontEndLib.js";

	// post request to server Async
	this.sendPostRequest_Async = function(page, params, callbackf) {
		var tmpXMLhttp = null;
		params.token = this.token;
		if (window.XMLHttpRequest) {
			// code for IE7+, Firefox, Chrome, Opera, Safari
			tmpXMLhttp = new window.XMLHttpRequest();
		};
		tmpXMLhttp.onreadystatechange=function() {
			if (tmpXMLhttp.readyState==4 && tmpXMLhttp.status==200) {
				if(tmpXMLhttp.responseText == "")
					obj = { "result" : "fail" };
				else
				{
					alert(tmpXMLhttp.responseText);
					var obj = JSON.parse(tmpXMLhttp.responseText);
					callbackf(obj);
					delete obj;
					delete tmpXMLhttp;
				}
			}
		}
		tmpXMLhttp.open("POST", this.baseUrl + page, true);
		tmpXMLhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
		tmpXMLhttp.send(this.createUrlFromObj(params));
	};

	// post request to server Sync
	this.sendPostRequest_Sync = function(page, params) {
		params.token = this.token;
		// alert(this.createUrlFromObj(params));
		
		var tmpXMLhttp = null;
		
		var obj = null;
		if (window.XMLHttpRequest) {
			// code for IE7+, Firefox, Chrome, Opera, Safari
			tmpXMLhttp = new window.XMLHttpRequest();
		};
		tmpXMLhttp.onreadystatechange=function() {
			if (tmpXMLhttp.readyState==4 && tmpXMLhttp.status==200) {
				if(tmpXMLhttp.responseText == "")
					obj = { "result" : "fail" };
				else
				{
					obj = JSON.parse(tmpXMLhttp.responseText);
					delete tmpXMLhttp;
				}
			}
		}
		tmpXMLhttp.open("POST", this.baseUrl + page, false);
		tmpXMLhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
		tmpXMLhttp.send(this.createUrlFromObj(params));
		return obj;
	};

	this.auth = new (function(t) {
		this.p = t;
		this.sign_in = function (email, password) {
			var params = {};
			params.email = email;
			params.password = password;
			params.client = this.p.client;
			var obj = this.p.sendPostRequest_Sync('api/auth/sign_in.php', params);
			// alert(JSON.stringify(obj));
			var bRes = obj.result == "ok";
			if (bRes)
				this.p.token = obj.token;
			else
				this.p.token = "";
			return bRes;
		};
		this.sign_out = function () {
			var params = {};
			var obj = this.p.sendPostRequest_Sync('api/auth/sign_out.php', params);
			this.p.token = "";
			return true;
		};
	})(this);

	
	this.games = new (function(t) {
		this.p = t;
		this.list = function() {
			var params = {};
			var obj = this.p.sendPostRequest_Sync('api/games/list.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
		this.get = function(gameid) {
			var params = {};
			params.id = gameid;
			var obj = this.p.sendPostRequest_Sync('api/games/get.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
		this.choose = function(gameid) {
			var params = {};
			params.id = gameid;
			var obj = this.p.sendPostRequest_Sync('api/games/choose.php', params);
			// alert(JSON.stringify(params));
			var bRes = obj.result == "ok";
			return bRes ? obj : obj.error;
		};
	})(this);
	
	this.quests = new (function(t) {
		this.p = t;
		this.list = function() {
			var params = {};
			var obj = this.p.sendPostRequest_Sync('api/quests/list.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj : obj.error;
		};
		this.get = function(questid) {
			var params = {};
			params.taskid = questid;
			var obj = this.p.sendPostRequest_Sync('api/quests/get.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    // return all information if you are admin
  	this.get_all = function(questid) {
			var params = {};
			params.questid = questid;
			var obj = this.p.sendPostRequest_Sync('api/quests/get_all.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    // remove quest if you are admin
    this.delete = function(questid) {
			var params = {};
			params.questid = questid;
			var obj = this.p.sendPostRequest_Sync('api/quests/delete.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    this.take = function(questid) {
			var params = {};
			params.questid = questid;
			var obj = this.p.sendPostRequest_Sync('api/quests/take.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    this.pass = function(questid, answer) {
			var params = {};
			params.questid = questid;
      params.answer = answer;
			var obj = this.p.sendPostRequest_Sync('api/quests/pass.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    // insert quest if you are admin
    this.insert = function(quest_uuid, name, short_text, text, score, min_score, subject, idauthor, author, answer, state, description_state) {
			var params = {};
			params.quest_uuid = quest_uuid;
      params.name = name;
      params.short_text = short_text;
      params.text = text;
      params.score = score;
      params.min_score = min_score;
      params.subject = subject;
      params.idauthor = idauthor;
      params.author = author;
      params.answer = answer;
      params.state = state;
      params.description_state = description_state;
			var obj = this.p.sendPostRequest_Sync('api/quests/insert.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    // update quest info if you are admin
    this.update = function(questid, name, short_text, text, score, min_score, subject, idauthor, author, answer, state, description_state) {
			var params = {};
			params.questid = questid;
      params.name = name;
      params.short_text = short_text;
      params.text = text;
      params.score = score;
      params.min_score = min_score;
      params.subject = subject;
      params.idauthor = idauthor;
      params.author = author;
      params.answer = answer;
      params.state = state;
      params.description_state = description_state;
			var obj = this.p.sendPostRequest_Sync('api/quests/update.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
    // update quest info if you are admin
    this.user_answers = function(questid) {
			var params = {};
			params.questid = questid;
			var obj = this.p.sendPostRequest_Sync('api/quests/user_answers.php', params);
			var bRes = obj.result == "ok";
			return bRes ? obj.data : obj.error;
		};
	})(this);
};

File: frontend/python-client/FHQFrontEndLib.py
#!/usr/bin/python

import requests
import uuid 

# api for work security
class FHQSecurity:
	def  __init__(self, parent):
		self.parent = parent
	def login(self, email, password):
		resp = self.parent.sendrequest('/security/login.php', {"email": email, "password": password})
		if resp['result'] == 'fail':
			return False
		self.parent.token = resp['data']['token']
		return True
	# def logoff(self):
	# register
	# restore password

# api for work with games
class FHQGames:
	def  __init__(self, parent):
		self.parent = parent
	
	def list(self):
		resp = self.parent.sendrequest('/games/list.php', {})
		return resp
		
	def choose(self, gameid):
		resp = self.parent.sendrequest('/games/choose.php', {'id' : gameid})
		return resp

	def get(self, gameid):
		resp = self.parent.sendrequest('/games/get.php', {'id' : gameid})
		return resp

	def insert(self, params):
		resp = self.parent.sendrequest('/games/insert.php', params)
		return resp

	def update(self, params):
		resp = self.parent.sendrequest('/games/update.php', params)
		return resp

	def delete(self, gameid):
		resp = self.parent.sendrequest('/games/delete.php', {'id' : gameid})
		return resp

	def update_rules(self, gameid, rules):
		resp = self.parent.sendrequest('/games/update_rules.php', {'id' : gameid, 'rules' : rules})
		return resp
		
	def update_score(self, gameid):
		resp = self.parent.sendrequest('/games/update_score.php', {'gameid' : gameid})
		return resp

	# def scoreboard
	# def upload_logo

class FHQQuests:
	def  __init__(self, parent):
		self.parent = parent

	def list(self, params):
		resp = self.parent.sendrequest('/quests/list.php', params)
		return resp
		
	def insert(self, params):
		resp = self.parent.sendrequest('/quests/insert.php', params)
		return resp

	def get(self, questid):
		resp = self.parent.sendrequest('/quests/get.php', { 'taskid' : questid })
		return resp

	def take(self, questid):
		resp = self.parent.sendrequest('/quests/take.php', { 'questid' : questid})
		return resp
		
	def trypass(self, questid, answer):
		resp = self.parent.sendrequest('/quests/pass.php', { 'questid' : questid, 'answer' : answer })
		return resp
		
	def delete(self, questid):
		resp = self.parent.sendrequest('/quests/delete.php', { 'questid' : questid })
		return resp

class FHQFrontEndLib:
	def __init__(self, url):
		self.token = ""
		self.url = url
		self.gameid = 0
		self.security = FHQSecurity(self)
		self.games = FHQGames(self)
		self.quests = FHQQuests(self)
		
	def sendrequest(self, path, params):
		# token
		if self.token != '':
			params['token'] = self.token

		# print params
		# todo: try catch	
		resp = {'result' : 'fail'}
		r = requests.post(self.url + path, params)
		if r.headers['content-type'] == 'application/json':
			try:
				resp = r.json()
			except ValueError:
				print r.text
				raise Exception('invalid json')

		if resp['result'] == 'fail':
			print ' *** FAIL *** '
			print ' * URL: ' + self.url + path
			print ' * Response:' + r.text
			print(' * Error.Code: ' + str(resp['error']['code']))
			print(' * Error.Message: ' + str(resp['error']['message']))
			print ' *************'
			return resp
		return resp
